File: nudge-generation/scripts/lib/prompt_constants_loader.py
"""
This source file is part of the Stanford Biodesign Digital Health MyHeart Counts open-source project based on the Stanford Spezi Template Application project

SPDX-FileCopyrightText: 2025-2026 Stanford University and the project authors (see CONTRIBUTORS.md)

SPDX-License-Identifier: MIT
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


PROMPT_SNIPPET_CATEGORIES = (
    "genderIdentity",
    "ageGroup",
    "disease",
    "stageOfChange",
    "educationLevel",
    "language",
)


def _default_prompt_constants_path() -> Path:
    return Path(__file__).resolve().parents[2] / "config/prompts/prompt_constants.v1.json"


def _default_prompt_schema_path() -> Path:
    return Path(__file__).resolve().parents[2] / "config/prompts/prompt_constants.schema.json"


def _require_string(value: Any, field: str) -> str:
    if not isinstance(value, str):
        raise ValueError(f"{field} must be a string")
    return value


def _require_string_map(value: Any, field: str) -> dict[str, str]:
    if not isinstance(value, dict):
        raise ValueError(f"{field} must be an object")
    output: dict[str, str] = {}
    for key, item in value.items():
        if not isinstance(key, str):
            raise ValueError(f"{field} contains a non-string key")
        if not isinstance(item, str):
            raise ValueError(f"{field}.{key} must be a string")
        output[key] = item
    return output


def validate_prompt_constants(payload: Any) -> dict[str, Any]:
    if not isinstance(payload, dict):
        raise ValueError("Prompt constants must be a JSON object")

    _require_string(payload.get("schemaVersion"), "schemaVersion")
    base_instruction = _require_string(
        payload.get("baseNudgeInstruction"), "baseNudgeInstruction"
    )
    if not base_instruction.strip():
        raise ValueError("baseNudgeInstruction must be non-empty")

    templates = payload.get("templates")
    if not isinstance(templates, dict):
        raise ValueError("templates must be an object")

    notification_template = _require_string(
        templates.get("notificationTimeContext"),
        "templates.notificationTimeContext",
    )
    if "{{preferredNotificationTime}}" not in notification_template:
        raise ValueError(
            "templates.notificationTimeContext must include {{preferredNotificationTime}}"
        )
    _require_string(templates.get("fullPromptAssembly"), "templates.fullPromptAssembly")

    context_snippets = payload.get("contextSnippets")
    if not isinstance(context_snippets, dict):
        raise ValueError("contextSnippets must be an object")

    defaults = payload.get("defaults")
    if not isinstance(defaults, dict):
        raise ValueError("defaults must be an object")

    for category in PROMPT_SNIPPET_CATEGORIES:
        _require_string_map(
            context_snippets.get(category), f"contextSnippets.{category}"
        )
        _require_string(defaults.get(category), f"defaults.{category}")

    return payload


def load_prompt_constants(path: str | Path | None = None) -> dict[str, Any]:
    constants_path = Path(path).resolve() if path else _default_prompt_constants_path()
    payload = json.loads(constants_path.read_text(encoding="utf-8"))
    validated = validate_prompt_constants(payload)

    # Parse schema file so failures surface early if the contract file is missing/invalid JSON.
    schema_path = _default_prompt_schema_path()
    if not schema_path.exists():
        raise ValueError(f"Prompt constants schema file not found: {schema_path}")
    json.loads(schema_path.read_text(encoding="utf-8"))
    return validated


def render_template(template: str, values: dict[str, str]) -> str:
    output = template
    for key, value in values.items():
        output = output.replace(f"{{{{{key}}}}}", value)
    return output


def get_context_snippet(
    constants: dict[str, Any], category: str, value: str | None
) -> str:
    if category not in PROMPT_SNIPPET_CATEGORIES:
        raise ValueError(f"Unsupported context snippet category: {category}")

    defaults = constants["defaults"]
    if not value:
        return defaults[category]

    snippets = constants["contextSnippets"][category]
    return snippets.get(value, defaults[category])


def build_prompt(context: dict[str, Any], path: str | Path | None = None) -> str:
    constants = load_prompt_constants(path)
    language_context = get_context_snippet(constants, "language", context["language"])
    gender_context = get_context_snippet(
        constants, "genderIdentity", context["genderIdentity"]
    )
    age_context = get_context_snippet(constants, "ageGroup", context["ageGroup"])
    disease_context = get_context_snippet(constants, "disease", context.get("disease"))
    stage_context = get_context_snippet(
        constants, "stageOfChange", context.get("stageOfChange")
    )
    education_context = get_context_snippet(
        constants, "educationLevel", context["educationLevel"]
    )
    notification_time_context = render_template(
        constants["templates"]["notificationTimeContext"],
        {"preferredNotificationTime": context["preferredNotificationTime"]},
    )

    return render_template(
        constants["templates"]["fullPromptAssembly"],
        {
            "baseNudgeInstruction": constants["baseNudgeInstruction"],
            "languageContext": language_context,
            "genderContext": gender_context,
            "ageContext": age_context,
            "diseaseContext": disease_context,
            "stageContext": stage_context,
            "educationContext": education_context,
            "notificationTimeContext": notification_time_context,
        },
    )

File: nudge-generation/src/config/promptConstantsLoader.ts
//
// This source file is part of the Stanford Biodesign Digital Health MyHeart Counts open-source project based on the Stanford Spezi Template Application project
//
// SPDX-FileCopyrightText: 2025-2026 Stanford University and the project authors (see CONTRIBUTORS.md)
//
// SPDX-License-Identifier: MIT
//

import * as fs from "fs";
import * as path from "path";
import { fileURLToPath } from "url";

export interface PromptContextInput {
  genderIdentity: string;
  ageGroup: string;
  disease: string | null;
  stageOfChange: string | null;
  educationLevel: string;
  language: string;
  preferredNotificationTime: string;
}

type PromptSnippetCategory =
  | "genderIdentity"
  | "ageGroup"
  | "disease"
  | "stageOfChange"
  | "educationLevel"
  | "language";

interface PromptConstants {
  schemaVersion: string;
  baseNudgeInstruction: string;
  templates: {
    notificationTimeContext: string;
    fullPromptAssembly: string;
  };
  contextSnippets: Record<PromptSnippetCategory, Record<string, string>>;
  defaults: Record<PromptSnippetCategory, string>;
}

let cachedPromptConstants: PromptConstants | null = null;
let checkedPromptSchema = false;

const getDefaultPromptConstantsPath = (): string => {
  const __filename = fileURLToPath(import.meta.url);
  const __dirname = path.dirname(__filename);
  return path.resolve(
    __dirname,
    "../../config/prompts/prompt_constants.v1.json",
  );
};

const getDefaultPromptSchemaPath = (): string => {
  const __filename = fileURLToPath(import.meta.url);
  const __dirname = path.dirname(__filename);
  return path.resolve(
    __dirname,
    "../../config/prompts/prompt_constants.schema.json",
  );
};

const isRecord = (value: unknown): value is Record<string, unknown> =>
  value !== null && typeof value === "object" && !Array.isArray(value);

const getStringRecord = (
  value: unknown,
  errorPrefix: string,
): Record<string, string> => {
  if (!isRecord(value)) {
    throw new Error(`${errorPrefix} must be an object`);
  }
  const output: Record<string, string> = {};
  for (const [key, item] of Object.entries(value)) {
    if (typeof item !== "string") {
      throw new Error(`${errorPrefix}.${key} must be a string`);
    }
    output[key] = item;
  }
  return output;
};

const validatePromptConstants = (
  raw: unknown,
  sourcePath: string,
): PromptConstants => {
  if (!isRecord(raw)) {
    throw new Error(`Prompt constants at ${sourcePath} must be a JSON object`);
  }

  if (
    typeof raw.schemaVersion !== "string" ||
    raw.schemaVersion.trim().length === 0
  ) {
    throw new Error(`schemaVersion is missing or invalid in ${sourcePath}`);
  }
  if (
    typeof raw.baseNudgeInstruction !== "string" ||
    raw.baseNudgeInstruction.trim().length === 0
  ) {
    throw new Error(
      `baseNudgeInstruction is missing or invalid in ${sourcePath}`,
    );
  }
  if (!isRecord(raw.templates)) {
    throw new Error(`templates is missing or invalid in ${sourcePath}`);
  }
  if (typeof raw.templates.notificationTimeContext !== "string") {
    throw new Error(`templates.notificationTimeContext must be a string`);
  }
  if (
    !raw.templates.notificationTimeContext.includes(
      "{{preferredNotificationTime}}",
    )
  ) {
    throw new Error(
      "templates.notificationTimeContext must include {{preferredNotificationTime}}",
    );
  }
  if (typeof raw.templates.fullPromptAssembly !== "string") {
    throw new Error(`templates.fullPromptAssembly must be a string`);
  }

  if (!isRecord(raw.contextSnippets)) {
    throw new Error(`contextSnippets is missing or invalid in ${sourcePath}`);
  }
  if (!isRecord(raw.defaults)) {
    throw new Error(`defaults is missing or invalid in ${sourcePath}`);
  }

  const categories: PromptSnippetCategory[] = [
    "genderIdentity",
    "ageGroup",
    "disease",
    "stageOfChange",
    "educationLevel",
    "language",
  ];

  const contextSnippets = {} as Record<
    PromptSnippetCategory,
    Record<string, string>
  >;
  const defaults = {} as Record<PromptSnippetCategory, string>;

  for (const category of categories) {
    contextSnippets[category] = getStringRecord(
      raw.contextSnippets[category],
      `contextSnippets.${category}`,
    );
    const defaultValue = raw.defaults[category];
    if (typeof defaultValue !== "string") {
      throw new Error(`defaults.${category} must be a string`);
    }
    defaults[category] = defaultValue;
  }

  return {
    schemaVersion: raw.schemaVersion,
    baseNudgeInstruction: raw.baseNudgeInstruction,
    templates: {
      notificationTimeContext: raw.templates.notificationTimeContext,
      fullPromptAssembly: raw.templates.fullPromptAssembly,
    },
    contextSnippets,
    defaults,
  };
};

export const loadPromptConstants = (filePath?: string): PromptConstants => {
  if (!filePath && cachedPromptConstants) {
    return cachedPromptConstants;
  }

  const resolvedPath = path.resolve(
    filePath ?? getDefaultPromptConstantsPath(),
  );
  const parsed = JSON.parse(fs.readFileSync(resolvedPath, "utf8")) as unknown;
  const validated = validatePromptConstants(parsed, resolvedPath);

  if (!checkedPromptSchema) {
    const schemaPath = getDefaultPromptSchemaPath();
    const schema = JSON.parse(fs.readFileSync(schemaPath, "utf8")) as unknown;
    if (!isRecord(schema) || typeof schema.$schema !== "string") {
      throw new Error(`Prompt constants schema is invalid at ${schemaPath}`);
    }
    checkedPromptSchema = true;
  }

  if (!filePath) {
    cachedPromptConstants = validated;
  }

  return validated;
};

export const renderTemplate = (
  template: string,
  values: Record<string, string>,
): string =>
  template.replace(/\{\{(\w+)\}\}/g, (_, key: string) => values[key] ?? "");

export const getContextSnippet = (
  constants: PromptConstants,
  category: PromptSnippetCategory,
  key: string | null,
): string => {
  if (!key) {
    return constants.defaults[category];
  }
  return (
    constants.contextSnippets[category][key] ?? constants.defaults[category]
  );
};

export const buildPrompt = (
  context: PromptContextInput,
  filePath?: string,
): string => {
  const constants = loadPromptConstants(filePath);
  const languageContext = getContextSnippet(
    constants,
    "language",
    context.language,
  );
  const genderContext = getContextSnippet(
    constants,
    "genderIdentity",
    context.genderIdentity,
  );
  const ageContext = getContextSnippet(constants, "ageGroup", context.ageGroup);
  const diseaseContext = getContextSnippet(
    constants,
    "disease",
    context.disease,
  );
  const stageContext = getContextSnippet(
    constants,
    "stageOfChange",
    context.stageOfChange,
  );
  const educationContext = getContextSnippet(
    constants,
    "educationLevel",
    context.educationLevel,
  );
  const notificationTimeContext = renderTemplate(
    constants.templates.notificationTimeContext,
    {
      preferredNotificationTime: context.preferredNotificationTime,
    },
  );

  return renderTemplate(constants.templates.fullPromptAssembly, {
    baseNudgeInstruction: constants.baseNudgeInstruction,
    languageContext,
    genderContext,
    ageContext,
    diseaseContext,
    stageContext,
    educationContext,
    notificationTimeContext,
  });
};
